### PR TITLE
fix: Correctly verify signatures when targetRevision is a branch name

### LINKED ISSUE #14009: verifyGnuPGSignature check triggers default case (Could not verify commit signature on revision)
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

gitlab and argocd UI say GPG signature is valid but sync errors with:

ComparisonError - Could not verify commit signature on revision '910788d9d64dcb613ba1c1c2b21aeeb07121ca91', check logs for more information.

log does not provide more info.

looks like the code is ending up in this default case here:
https://github.com/argoproj/argo-cd/blob/aa3b95561b0ad6902d7292819e2233257e6a2717/controller/state.go#L343-L345

**To Reproduce**

use git with GPG signature and enforce signature

**Expected behavior**

GPG check is working correct like in V2.5.18 or V2.6.8

**Screenshots**

image

image

image

**Version**

2.6.9 shows the same problem.

deployed via https://raw.githubusercontent.com/argoproj/argo-cd/v2.7.4/manifests/install.yaml

```shell
argocd: v2.7.4+a33baa3.dirty
  BuildDate: 2023-06-05T19:00:34Z
  GitCommit: a33baa301fe61b899dc8bbad9e554efbc77e0991
  GitTreeState: dirty
  GoVersion: go1.19.6
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

```
argocd-repo-server-59c98499f8-wt2qx argocd-repo-server time="2023-06-12T19:00:40Z" level=info msg="manifest cache hit: &ApplicationSource{RepoURL:https://gitlab.xxxxxxxx/k8s/gitops_k8s_local.git,Path:kube-system/metallb,TargetRevision:master,Helm:nil,Kustomize:nil,Directory:nil,Plugin:nil,Chart:,Ref:,}/910788d9d64dcb613ba1c1c2b21aeeb07121ca91"
argocd-repo-server-59c98499f8-wt2qx argocd-repo-server time="2023-06-12T19:00:40Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-06-12T19:00:40Z" grpc.time_ms=3.251 span.kind=server system=grpc
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="getRepoObjs stats" application=argocd/kube-system-metallb build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=24 unmarshal_ms=24 version_ms=0
argocd-repo-server-59c98499f8-wt2qx argocd-repo-server time="2023-06-12T19:00:40Z" level=info msg="manifest cache hit: &ApplicationSource{RepoURL:https://gitlab.xxxxxxxx/k8s/gitops_k8s_local.git,Path:kube-system,TargetRevision:master,Helm:nil,Kustomize:nil,Directory:nil,Plugin:nil,Chart:,Ref:,}/910788d9d64dcb613ba1c1c2b21aeeb07121ca91"
argocd-repo-server-59c98499f8-wt2qx argocd-repo-server time="2023-06-12T19:00:40Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2023-06-12T19:00:40Z" grpc.time_ms=0.879 span.kind=server system=grpc
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="getRepoObjs stats" application=argocd/kube-system build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=10 unmarshal_ms=10 version_ms=0
argocd-server-65d6c845b6-bc688 argocd-server time="2023-06-12T19:00:40Z" level=info msg="received unary call /application.ApplicationService/GetApplicationSyncWindows" grpc.method=GetApplicationSyncWindows grpc.request.content="name:\"kube-system-metallb\" appNamespace:\"argocd\" " grpc.service=application.ApplicationService grpc.start_time="2023-06-12T19:00:40Z" span.kind=server system=grpc
argocd-server-65d6c845b6-bc688 argocd-server time="2023-06-12T19:00:40Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=GetApplicationSyncWindows grpc.service=application.ApplicationService grpc.start_time="2023-06-12T19:00:40Z" grpc.time_ms=17.355 span.kind=server system=grpc
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="No status changes. Skipping patch" application=argocd/kube-system
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="Reconciliation completed" application=argocd/kube-system dedup_ms=0 dest-name= dest-namespace=kube-system dest-server="https://kubernetes.default.svc" diff_ms=19 fields.level=1 git_ms=10 health_ms=0 live_ms=0 settings_ms=0 sync_ms=0 time_ms=41
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="updated 'argocd/kube-system-metallb' operation (phase: Error)"
argocd-application-controller-0 argocd-application-controller time="2023-06-12T19:00:40Z" level=info msg="Sync operation to 910788d9d64dcb613ba1c1c2b21aeeb07121ca91 failed: ComparisonError: Could not verify commit signature on revision '910788d9d64dcb613ba1c1c2b21aeeb07121ca91', check logs for more information." application=kube-system-metallb dest-namespace=kube-system dest-server="https://kubernetes.default.svc" reason=OperationCompleted type=Warning
argocd-notifications-controller-65cddcf9d6-hw5wk argocd-notifications-controller time="2023-06-12T19:00:40Z" level=info msg="Start processing" resource=argocd/kube-system-metallb
argocd-notifications-controller-65cddcf9d6-hw5wk argocd-notifications-controller time="2023-06-12T19:00:40Z" level=info msg="Processing skipped: sync status out of date" resource=argocd/kube-system-metallb
```

example gitlog:

```
$ git log
commit 507df20b0167191986e6cfdd3f096557ab99b67c (HEAD -> master, origin/master, origin/HEAD)
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 19:37:30 2023 +0000

    release argocd based on pipelines/583 - d0d786ff92449cd7c88216ddc90ff7983a802eec

commit 9ad88216630761dca48e19149713944165c8ea53
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 19:29:00 2023 +0000

    release argocd based on pipelines/582 - 3e183275cac2914e41386cf3df8311379a160260

commit 0bd283decc59ddab5a6bdd1fe4a7f293fd7844ab
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 19:23:33 2023 +0000

    release argocd based on pipelines/581 - 7e36a50f366a90bced4feb1ad1b3fd6af408b373

commit 474fbd25630f7ef184ac1a500434c1dea9e18360
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 19:15:00 2023 +0000

    release argocd based on pipelines/580 - 896086e1694aa0633cec0afe23a52c91e2405808

commit 910788d9d64dcb613ba1c1c2b21aeeb07121ca91
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 18:52:30 2023 +0000

    release argocd based on pipelines/579 - 3a351cfe25bbddb0e3d3997d5451631524d1eb54

commit da4fdd6f58267efb324600c438606aa336403620
Author: gitlab_ci_user <gitlab@XXXXXXXXXX>
Date:   Mon Jun 12 18:47:34 2023 +0000

    release argocd based on pipelines/578 - 88b2449b6a878f31676906870b9508ecce232a97
```

### PR DESCRIPTION
**Pull request by jannfis**
Fixes #14009 

During manifest rendering, Argo CD checks out the resolved SHA of the `targetRevision`. This leads the repository into a detached HEAD state, from where the branch name is not reachable.

We do need to pass a tag name (i.e. the unresolved `targetRevision`) when validating signatures on a signed (annotated) tag though. So we'll have the repository server figure out whether a `targetRevision` is specifying an annotated tag or not, and call the verification routines accordingly.

Technically, we do not need `hack/git-verify-wrapper.sh` anymore. However, I'm going to replace the wrapper by Go code in a future PR, because this particular PR fixes a regression and needs to be cherry-picked.

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: util/git/client.go [hunk 1]
```diff
@@ -653,17 +658,28 @@ func (m *nativeGitClient) VerifyCommitSignature(revision string) (string, error)
 	return out, nil
 }
 
+// IsAnnotatedTag returns true if the revision points to an annotated tag
+func (m *nativeGitClient) IsAnnotatedTag(revision string) bool {
```
Comment (crenshaw-dev) [inline util/git/client.go:662]: Should we unit test this function individually? Seems like a fragile spot.

Comment (jannfis) [reply util/git/client.go:662]: It's covered by the end-to-end tests, but I agree that a unit test would be helpful.

Comment (jannfis) [reply util/git/client.go:663]: Added unit-tests, caught a bug :grin:

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**